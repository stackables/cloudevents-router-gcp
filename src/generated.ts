import { LogEntryData } from '@google/events/cloud/audit/v1/LogEntryData'
import { BuildEventData } from '@google/events/cloud/cloudbuild/v1/BuildEventData'
import { DocumentEventData } from '@google/events/cloud/firestore/v1/DocumentEventData'
import { MessagePublishedData } from '@google/events/cloud/pubsub/v1/MessagePublishedData'
import { SchedulerJobData } from '@google/events/cloud/scheduler/v1/SchedulerJobData'
import { StorageObjectData } from '@google/events/cloud/storage/v1/StorageObjectData'
import { AnalyticsLogData } from '@google/events/firebase/analytics/v1/AnalyticsLogData'
import { AuthEventData } from '@google/events/firebase/auth/v1/AuthEventData'
import { ReferenceEventData } from '@google/events/firebase/database/v1/ReferenceEventData'
import { RemoteConfigEventData } from '@google/events/firebase/remoteconfig/v1/RemoteConfigEventData'

export type GoogleEvents = {
  'google.cloud.audit.log.v1.written': LogEntryData
  'google.cloud.cloudbuild.build.v1.statusChanged': BuildEventData
  'google.cloud.firestore.document.v1.created': DocumentEventData
  'google.cloud.firestore.document.v1.updated': DocumentEventData
  'google.cloud.firestore.document.v1.deleted': DocumentEventData
  'google.cloud.firestore.document.v1.written': DocumentEventData
  'google.cloud.pubsub.topic.v1.messagePublished': MessagePublishedData
  'google.cloud.scheduler.job.v1.executed': SchedulerJobData
  'google.cloud.storage.object.v1.finalized': StorageObjectData
  'google.cloud.storage.object.v1.archived': StorageObjectData
  'google.cloud.storage.object.v1.deleted': StorageObjectData
  'google.cloud.storage.object.v1.metadataUpdated': StorageObjectData
  'google.firebase.analytics.log.v1.written': AnalyticsLogData
  'google.firebase.auth.user.v1.created': AuthEventData
  'google.firebase.auth.user.v1.deleted': AuthEventData
  'google.firebase.database.ref.v1.created': ReferenceEventData
  'google.firebase.database.ref.v1.updated': ReferenceEventData
  'google.firebase.database.ref.v1.deleted': ReferenceEventData
  'google.firebase.database.ref.v1.written': ReferenceEventData
  'google.firebase.remoteconfig.remoteConfig.v1.updated': RemoteConfigEventData
}