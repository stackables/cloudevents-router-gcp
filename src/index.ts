import type { Message, MessagePublishedData } from '@google/events/cloud/pubsub/v1/MessagePublishedData'

export type PubSubParsedMessage<T> =
    Omit<MessagePublishedData, "message"> & {
        message: Omit<Message, "data"> & {
            data: T
            messageId: string
        }
    }

export { GoogleEvents } from './generated'
export { republishPubSubByTopic } from './pubsub-splitter'