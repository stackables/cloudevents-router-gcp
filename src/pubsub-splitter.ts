import type { CloudEventsRouter } from 'cloudevents-router'
import { GoogleEvents } from '.'

type PublishOptions<T> = {
    topics: { [key: string]: keyof T }
    onUnhandled?: 'DEFAULT' | 'REJECT' | 'IGNORE'
}

export function republishPubSubByTopic<T extends GoogleEvents>(events: CloudEventsRouter<T>, opts: PublishOptions<T>) {

    events.on('google.cloud.pubsub.topic.v1.messagePublished', (event) => {

        // Topic format "//pubsub.googleapis.com/projects/serverless-com-demo/topics/my-topic"
        const topic = event.source.split('/').pop() as string
        const mapped = opts.topics?.[topic] as string

        if (!mapped && opts.onUnhandled === 'REJECT') {
            throw new Error('Message from unknown topic: ' + topic)
        }

        if (!mapped && opts.onUnhandled === 'IGNORE') {
            return
        }

        // Republish event
        const e = event.cloneWith({
            type: mapped,
            data: {
                ...event.data,
                message: {
                    ...event.data?.message,
                    data: event.data?.message?.data ? JSON.parse(Buffer.from(event.data.message.data, 'base64').toString()) : undefined
                }
            }
        })

        return events.process(e)
    })

}