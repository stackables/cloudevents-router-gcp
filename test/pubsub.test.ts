import { CloudEvent } from 'cloudevents'
import { CloudEventsRouter } from 'cloudevents-router'
import { GoogleEvents, PubSubParsedMessage, republishPubSubByTopic } from '../src'
import omitDeep from 'omit-deep-lodash'

// Define message payload type
type EventMap = GoogleEvents & {
    'pubsub.userAdded': PubSubParsedMessage<{
        username: string
        password: string
        age?: number
    }>
    'pubsub.profileChanged': PubSubParsedMessage<{
        username: string
        profilePicture: string
    }>
    'pubsub.emptyData': PubSubParsedMessage<void>
}

function publishFakeEvent(router: CloudEventsRouter<any>, topic: string, payload: any) {

    const ce = new CloudEvent({
        type: 'google.cloud.pubsub.topic.v1.messagePublished',
        source: `//pubsub.googleapis.com/projects/test-project-id/topics/${topic}`,
        data: {
            message: {
                data: payload ? Buffer.from(JSON.stringify(payload)).toString('base64') : undefined,
                messageId: 'my-message-id',
                publishTime: '2020-08-14T20:50:04.994Z',
            },
            subscription: 'subscription',
        }
    })

    return router.process(ce)
}

test('Republish PubSub example - default', async () => {

    const events: any[] = []
    const router = new CloudEventsRouter<EventMap>()

    router.on('pubsub.profileChanged', async (event) => {
        events.push(event)
        events.push(event.data)
    })

    router.onUnhandled(async (event) => {
        events.push(event)
        events.push(event.data)
    })

    // republish the event
    republishPubSubByTopic(router, {
        topics: {
            'user-added-topic': 'pubsub.userAdded',
            'profile-changed-topic': 'pubsub.profileChanged'
        },
    })

    await publishFakeEvent(router, 'profile-changed-topic', { username: 'testuser1', profilePicture: 'http://picture.com' })
    await publishFakeEvent(router, 'user-added-topic', { username: 'testuser1', password: '***' })
    expect(omitDeep(events, 'id', 'time')).toMatchSnapshot()
})

test('Republish PubSub example - ignore', async () => {

    const events: any[] = []
    const router = new CloudEventsRouter<EventMap>()

    router.on('pubsub.profileChanged', async (event) => {
        events.push(event)
        events.push(event.data)
    })

    router.onUnhandled(async (event) => {
        events.push(event)
        events.push(event.data)
    })

    // republish the event
    republishPubSubByTopic(router, {
        topics: {
            'profile-changed-topic': 'pubsub.profileChanged'
        },
        onUnhandled: 'IGNORE'
    })

    await publishFakeEvent(router, 'profile-changed-topic', { username: 'testuser1', profilePicture: 'http://picture.com' })
    await publishFakeEvent(router, 'user-added-topic', { username: 'testuser1', password: '***' })
    expect(omitDeep(events, 'id', 'time')).toMatchSnapshot()
})

test('Republish PubSub example - reject', async () => {

    const events: any[] = []
    const router = new CloudEventsRouter<EventMap>()

    router.on('pubsub.profileChanged', async (event) => {
        events.push(event)
        events.push(event.data)
    })

    router.onUnhandled(async (event) => {
        events.push(event)
        events.push(event.data)
    })

    // republish the event
    republishPubSubByTopic(router, {
        topics: {
            'profile-changed-topic': 'pubsub.profileChanged'
        },
        onUnhandled: 'REJECT'
    })

    await publishFakeEvent(router, 'profile-changed-topic', { username: 'testuser1', profilePicture: 'http://picture.com' })
    await expect(() => publishFakeEvent(router, 'user-added-topic', { username: 'testuser1', password: '***' })).rejects.toThrow()
    expect(omitDeep(events, 'id', 'time')).toMatchSnapshot()
})

test('Republish PubSub example - empty', async () => {

    const events: any[] = []
    const router = new CloudEventsRouter<EventMap>()

    router.onUnhandled(async (event) => {
        events.push(event)
        events.push(event.data)
    })

    // republish the event
    republishPubSubByTopic(router, {
        topics: {
            'empty-message-test': 'pubsub.emptyData'
        }
    })

    await publishFakeEvent(router, 'empty-message-test', undefined)
    expect(omitDeep(events, 'id', 'time')).toMatchSnapshot()
})
