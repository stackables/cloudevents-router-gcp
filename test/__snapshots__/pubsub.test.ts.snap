// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Republish PubSub example - default 1`] = `
Array [
  Object {
    "data_base64": undefined,
    "datacontentencoding": undefined,
    "datacontenttype": undefined,
    "dataschema": undefined,
    "schemaurl": undefined,
    "source": "//pubsub.googleapis.com/projects/test-project-id/topics/profile-changed-topic",
    "specversion": "1.0",
    "subject": undefined,
    "type": "pubsub.profileChanged",
  },
  Object {
    "message": Object {
      "data": Object {
        "profilePicture": "http://picture.com",
        "username": "testuser1",
      },
      "messageId": "my-message-id",
      "publishTime": "2020-08-14T20:50:04.994Z",
    },
    "subscription": "subscription",
  },
  Object {
    "data_base64": undefined,
    "datacontentencoding": undefined,
    "datacontenttype": undefined,
    "dataschema": undefined,
    "schemaurl": undefined,
    "source": "//pubsub.googleapis.com/projects/test-project-id/topics/user-added-topic",
    "specversion": "1.0",
    "subject": undefined,
    "type": "pubsub.userAdded",
  },
  Object {
    "message": Object {
      "data": Object {
        "password": "***",
        "username": "testuser1",
      },
      "messageId": "my-message-id",
      "publishTime": "2020-08-14T20:50:04.994Z",
    },
    "subscription": "subscription",
  },
]
`;

exports[`Republish PubSub example - empty 1`] = `
Array [
  Object {
    "data_base64": undefined,
    "datacontentencoding": undefined,
    "datacontenttype": undefined,
    "dataschema": undefined,
    "schemaurl": undefined,
    "source": "//pubsub.googleapis.com/projects/test-project-id/topics/empty-message-test",
    "specversion": "1.0",
    "subject": undefined,
    "type": "pubsub.emptyData",
  },
  Object {
    "message": Object {
      "data": undefined,
      "messageId": "my-message-id",
      "publishTime": "2020-08-14T20:50:04.994Z",
    },
    "subscription": "subscription",
  },
]
`;

exports[`Republish PubSub example - ignore 1`] = `
Array [
  Object {
    "data_base64": undefined,
    "datacontentencoding": undefined,
    "datacontenttype": undefined,
    "dataschema": undefined,
    "schemaurl": undefined,
    "source": "//pubsub.googleapis.com/projects/test-project-id/topics/profile-changed-topic",
    "specversion": "1.0",
    "subject": undefined,
    "type": "pubsub.profileChanged",
  },
  Object {
    "message": Object {
      "data": Object {
        "profilePicture": "http://picture.com",
        "username": "testuser1",
      },
      "messageId": "my-message-id",
      "publishTime": "2020-08-14T20:50:04.994Z",
    },
    "subscription": "subscription",
  },
]
`;

exports[`Republish PubSub example - reject 1`] = `
Array [
  Object {
    "data_base64": undefined,
    "datacontentencoding": undefined,
    "datacontenttype": undefined,
    "dataschema": undefined,
    "schemaurl": undefined,
    "source": "//pubsub.googleapis.com/projects/test-project-id/topics/profile-changed-topic",
    "specversion": "1.0",
    "subject": undefined,
    "type": "pubsub.profileChanged",
  },
  Object {
    "message": Object {
      "data": Object {
        "profilePicture": "http://picture.com",
        "username": "testuser1",
      },
      "messageId": "my-message-id",
      "publishTime": "2020-08-14T20:50:04.994Z",
    },
    "subscription": "subscription",
  },
]
`;
